doc/sonic-application-extention/sonic-application-extention-hld.md

```diff
@@ -0,0 +1,1441 @@
+
+This section is WIP and describes the approach in very high level and needs more deep investigation.
+
+The first thing to note here is that a Kubernetes manifest file can be auto-generated from SONiC Package manifest as it cat provide all the info about
+how to run the container. This manifest auto-generation is related to Kubernetes master changes while we will be focusing on SONiC OS side, so it is out of
+the scope here.
+
+Besides of the manifest on the master, the SONiC switch has to have service files, configs, scripts installed for the feature.
+
+1. Package installed through SONiC package manager.
+
+During package installation these necessary components will be auto-generated and installed on the switch. The auto-generation must honor new requirements
+to support "local" and "kube" modes when in "local" mode the container gets started on systemctl start, while in "kube" mode the appropriate feature label
+is set.
+
+2. Docker container upgrades through Kubernetes.
+
+During that processes a new Docker container gets deployed and run on the switch. Kubernetes managed features introduces a state machine for a running
+container that is reflected in the STATE DB. When container starts it sets its state as "pending" and waits till there is a "ready" flag set.
+
+A state-db-watcherd is listening to those changes. If there is a need to regenerate service file and scripts it initiates sonic-package-manager
+upgrade re-generation processes. Since the image is already downloaded and running but pending, the package manifest can be read and based on manifest
+it can generate the required files. Then, the state-db-watcherd can proceed to systemctl stop old container, systemctl start new container
+where the container_action script set the container state to "ready" and the container resumes to run the applications.
+
+3. Docker container deploy through Kubernetes.
+
+The case describe the scenario when the user is setting a label via "config kube label add <key>=<value>". If labels match the manifest on the master
+the docker image gets deployed. Using the same approach, between "pending" and "ready" states an auto-generation process can happen and register the
+Docker as a new package. With that, a deployed image will become an installed SONiC Package, that can be managed in "local" mode as well.
+
+
+### 3rd party Docker images
+
+<!-- onit in toc -->
+### Installing 3rd party image as is.
+
+It is possible to install 3rd party Docker images if the installer will use a default manifest. This default manifest will have only mandatory fields
+and an auto-generated service will require only "docker" service to be started. The name of the service is derived from the name of the Package in *packages.json*.
+
+Default manifest:
+```json
+{
+  "service": {
+    "name": "<package-name>",
+    "requires": "docker",
+    "after": "docker"
```

Comment:
If the user wants to make use of the docker options say. cpu cores, memory..for a given docker, do we have any option in the manifest file?

Reply:
This is similar to the question with disk space limit. Having these constraints in manifest might not be a good idea because one switch may have powerful enough CPU and a lots of RAM while another switch platform might be much slower.
You may want to tweak this parameters differently for different platforms, but you can not put these parameters in manifest for every possible platform.
The plan is to have an option for the user to configure resource limits and to have something similar to above:
```
config feature cpu-limit
config feature memory-limit
etc.
```
But this is out of scope for this document.